### PR TITLE
Make manifest serialization deterministic

### LINKED ISSUE #4326: Manifest re-serialization in 'cargo package' is non-deterministic
In `Cargo.toml` files generated by `cargo vendor`, the order of dependencies seems not to be deterministic: when updating in Firefox a crate to a version that did not change the set of dependencies, I get a diff like below.

Could this be based on the iteration order of a `HashMap` that uses the default `RandomState` hasher? If so, what do you think of switching to a deterministic hasher? Or perhaps sorting before generating TOML files?

```diff
diff --git a/third_party/rust/cssparser/Cargo.toml b/third_party/rust/cssparser/Cargo.toml
index 76fadfd855e4..94e9f473274c 100644
--- a/third_party/rust/cssparser/Cargo.toml
+++ b/third_party/rust/cssparser/Cargo.toml
@@ -12,7 +12,7 @@
 
 [package]
 name = "cssparser"
-version = "0.18.0"
+version = "0.18.1"
 authors = ["Simon Sapin <simon.sapin@exyr.org>"]
 build = "build.rs"
 exclude = ["src/css-parsing-tests/**", "src/big-data-url.css"]
@@ -22,33 +22,33 @@ readme = "README.md"
 keywords = ["css", "syntax", "parser"]
 license = "MPL-2.0"
 repository = "https://github.com/servo/rust-cssparser"
-[dependencies.heapsize]
-version = ">= 0.3, < 0.5"
-optional = true
-
 [dependencies.serde]
 version = "1.0"
 optional = true
 
+[dependencies.heapsize]
+version = ">= 0.3, < 0.5"
+optional = true
+
 [dependencies.procedural-masquerade]
 version = "0.1"
 
-[dependencies.cssparser-macros]
-version = "0.3"
-
 [dependencies.matches]
 version = "0.1"
 
 [dependencies.phf]
 version = "0.7"
-[dev-dependencies.encoding_rs]
-version = "0.5"
+
+[dependencies.cssparser-macros]
+version = "0.3"
+[dev-dependencies.rustc-serialize]
+version = "0.3"
 
 [dev-dependencies.difference]
 version = "1.0"
 
-[dev-dependencies.rustc-serialize]
-version = "0.3"
+[dev-dependencies.encoding_rs]
+version = "0.5"
 [build-dependencies.syn]
 version = "0.11"
```

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes #4326

`cargo package` (and so `cargo publish`) parses a crate’s `Cargo.toml`, makes some modifications, and re-serializes it. Because the `TomlManifest` struct uses `HashMap` with its default `RandomState` hasher, the maps’ iteration order changed on every run.

As a result, when using `cargo vendor`, updating a dependency would generate a diff larger than necessary, with non-significant order-changes obscuring significant changes.

This replaces some uses of `HashMap` with `BTreeMap`, whose iteration order is deterministic (based on `Ord`).

### CODE REVIEW

**Comment by matklad** [comment]
Perhaps we should use `OrderedMap` here in some cases, to preserve the order of the original Cargo.toml?

**Comment by SimonSapin** [comment]
I personally don’t care much about preserving the original order, as long as it doesn’t change on every update.

File: Cargo.lock [hunk 1]
```diff
@@ -1028,11 +1055,14 @@ dependencies = [
...
 "checksum core-foundation-sys 0.4.4 (registry+https://github.com/rust-lang/crates.io-index)" = "bc9fb3d6cb663e6fd7cf1c63f9b144ee2b1e4a78595a0451dd34bff85b9a3387"
 "checksum crossbeam 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)" = "0c5ea215664ca264da8a9d9c3be80d2eaf30923c259d03e870388eb927508f97"
 "checksum crossbeam 0.3.0 (registry+https://github.com/rust-lang/crates.io-index)" = "8837ab96533202c5b610ed44bc7f4183e7957c1c8f56e8cc78bb098593c8ba0a"
+"checksum crypto-hash 0.3.0 (registry+https://github.com/rust-lang/crates.io-index)" = "34903878eec1694faf53cae8473a088df333181de421d4d3d48061d6559fe602"
 "checksum curl 0.4.8 (registry+https://github.com/rust-lang/crates.io-index)" = "7034c534a1d7d22f7971d6088aa9d281d219ef724026c3428092500f41ae9c2c"
 "checksum curl-sys 0.3.15 (registry+https://github.com/rust-lang/crates.io-index)" = "4bee31aa3a079d5f3ff9579ea4dcfb1b1a17a40886f5f467436d383e78134b55"
 "checksum custom_derive 0.1.7 (registry+https://github.com/rust-lang/crates.io-index)" = "ef8ae57c4978a2acd8b869ce6b9ca1dfe817bff704c220209fdef2c0b75a01b9"
```
Comment (matklad) [inline Cargo.lock]: Hm, where do these lockfile changes come from?

Comment (SimonSapin) [reply Cargo.lock]: They’re unrelated, and happen when I compile master unmodified. @alexcrichton said (in person) that it’s likely that someone forgot to include that change when they changed Cargo’s `Cargo.toml` in a previous PR, and that including that in this PR was fine. I can remove it if you prefer.

Comment (SimonSapin) [reply Cargo.lock]: Or split it into a separate commit in the same PR.

**Comment by matklad** [comment]
@bors r+


Thanks @SimonSapin !

**Pull request merged**